src/parser/jobs/sam/Gauge.js

```diff
@@ -0,0 +1,208 @@
+import {Suggestion, SEVERITY} from 'parser/core/modules/Suggestions'
+
+//future-proofing for more kenki actions
+
+const MAX_KENKI = 100
+
+const KENKI_BUILDERS = {
+
+	//single target
+	[ACTIONS.GEKKO.id]: 10,
+	[ACTIONS.KASHA.id]: 10,
+	[ACTIONS.YUKIKAZE.id]: 10,
+	[ACTIONS.HAKAZE.id]: 5,
+	[ACTIONS.JINPU.id]: 5,
+	[ACTIONS.SHIFU.id]: 5,
+	//aoe
+	[ACTIONS.MANGETSU.id]: 10,
+	[ACTIONS.OKA.id]: 10,
+	[ACTIONS.FUGA.id]: 5,
+	//ranged
+	[ACTIONS.ENPI.id]: 10,
+}
+
+const KENKI_SPENDERS = {
+	[ACTIONS.HISSATSU_GYOTEN.id]: 10,
+	[ACTIONS.HISSATSU_YATEN.id]: 10,
+	[ACTIONS.HISSATSU_SEIGAN.id]: 15,
+	[ACTIONS.HISSATSU_KAITEN.id]: 20,
+	[ACTIONS.HISSATSU_SHINTEN.id]: 25,
+	[ACTIONS.HISSATSU_KYUTEN.id]: 25,
+	[ACTIONS.HISSATSU_GUREN.id]: 50,
+}
+
+// sen stuff
+
+// 1 sen value for each finisher. using finisher with sen you already have is bad.
+const MAX_GEKKO_SEN = 1
+const MAX_KASHA_SEN = 1
+const MAX_YUKIKAZE_SEN = 1
+
+export default class Gauge extends Module {
+	static handle = 'gauge'
+	static dependencies = [
+		'gcd',
+		'combatants',
+		'cooldowns',
+		'suggestions',
+	]
+
+	//kenki
+	_kenki = 0
+	_wastedKenki = 0
+
+	//sen
+	_gekkosen = 0
+	_kashasen = 0
+	_yukikazesen = 0
+
+	_wastedsen = 0
+
+	constructor(...args) {
+		super(...args)
+		this.addHook('cast', {by: 'player'}, this._onCast)
+		this.addHook('death', {to: 'player'}, this._onDeath)
+		this.addHook('complete', this._onComplete)
+	}
+	//check for kenki value changes, then sen changes
+	_onCast(event) {
+		const abilityId = event.ability.guid
+		if (KENKI_BUILDERS[abilityId]) {
+			this._addKenki(abilityId)
+		}
+		if (KENKI_SPENDERS[abilityId]) {
+			this._kenki -= KENKI_SPENDERS[abilityId]
+		}
+		if (abilityId === ACTIONS.HAGAKURE.id) {
+			this._Sen2Kenki()
+		}
+		if (abilityId ===ACTIONS.HIGANBANA.id || abilityId === ACTIONS.TENKA_GOKEN.id || abilityId === ACTIONS.MIDARE_SETSUGEKKA.id) {
+			this._removeSen()
+		}
+		if (abilityId === ACTIONS.GEKKO.id || abilityId === ACTIONS.MANGETSU.id) {
+			this._addGekkoSen()
+		}
+		if (abilityId === ACTIONS.KASHA.id || abilityId === ACTIONS.OKA.id) {
+			this._addKashaSen()
+		}
+		if (abilityId === ACTIONS.YUKIKAZE.id) {
+			this._addYukikazeSen()
+		}
+
+	}
+
+	//kenki quick maths
+
+	_addKenki(abilityId) {
+		this._kenki += KENKI_BUILDERS[abilityId]
+		if (this._kenki > MAX_KENKI) {
+			const waste = this._kenki - MAX_KENKI
+			this._wastedKenki += waste
+			this._kenki = MAX_KENKI
+			return waste
+		}
+		return 0
+	}
+
+	_Sen2Kenki() {
+		this._kenki += ((this._gekkosen + this._kashasen + this._yukikazesen) * 20)
+		if (this.kenki > MAX_KENKI) {
+			const waste = this._kenki - MAX_KENKI
+			this._wastedKenki += waste
+			this._kenki = MAX_KENKI
+			return waste
+		}
+
+		this._gekkosen = 0
+		this._kashasen = 0
+		this._yukikazesen = 0
+
+		return 0
+	}
+
+	//sen calcs
+
+	_addGekkoSen() {
+		this._gekkosen += 1
+		if (this._gekkosen > MAX_GEKKO_SEN) {
+			const waste = this._gekkosen - MAX_GEKKO_SEN
```

Comment:
I think you'll want to cap `this._gekkosen` down to `MAX_GEKKO_SEN` after calculating waste here (and same goes for Kasha/Yukikaze). Otherwise, it'll just compound. Like, if someone used 3 Gekko finishers in a row, the first would be fine, the second would be +1 waste, and the third would be +2 waste. Also it'll throw off the Sen -> Kenki calculation, since you'd be getting 60 out of Gekko alone in this scenario.